api/ssl_certificates_service.go

```diff
@@ -0,0 +1,92 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+)
+
+type SSLCertificateInput struct {
+	CertPem       string `json:"certificate"`
+	PrivateKeyPem string `json:"private_key"`
+}
+
+type SSLCertificateOutput struct {
+	Certificate SSLCertificate `json:"ssl_certificate"`
+}
+
+type SSLCertificate struct {
+	Certificate string `json:"certificate"`
+}
+
+func (a Api) UpdateSSLCertificate(certBody SSLCertificateInput) error {
+	body, err := json.Marshal(certBody)
+	if err != nil {
+		return err // not tested
+	}
+
+	req, err := http.NewRequest("PUT", "/api/v0/settings/ssl_certificate", bytes.NewReader(body))
+	if err != nil {
+		return err // not tested
+	}
+	req.Header.Add("Content-Type", "application/json")
+
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if err = validateStatusOK(resp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (a Api) GetSSLCertificate() (SSLCertificateOutput, error) {
+	var output SSLCertificateOutput
+
+	req, err := http.NewRequest("GET", "/api/v0/settings/ssl_certificate", nil)
+	if err != nil {
+		return output, err
+	}
+
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return output, err
+	}
+
+	if err = validateStatusOK(resp); err != nil {
+		return SSLCertificateOutput{}, err
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&output)
+	if err != nil {
+		return output, err
+	}
+
+	if output.Certificate.Certificate == "" {
+		output.Certificate.Certificate = "Ops Manager Self Signed Cert"
```

Comment:
This definitely makes the output nicer when using the table format. But when using json, it is hard to be consumed programmatically, for two reasons:
1. the string is not actually valid cert, accidentally parsing it into x509 cert would cause an error if not caught.
1. it introduces a potential fragile dependency for the consumer. If the exact string match is used by the consumer, they need to rewrite the logic each time the hardcoded string changes (though not likely to change).

So what about moving the string out from `api` package? `GetSSLCertificate()` should respect the payload returned from opsman. And if we want a nice table output, we can have the hardcoded string added to the table at `ssl-certificate` command level. 

This way we can have the raw json if we use `-f json`, or a nice table if we use `-f table`.

Reply:
Thats a good suggestion, the raw json happens to be empty string in the case of an OpsMan self signed certificate. Would that be preferable for the json format and have a `if table then add Self signed cert string` case? In terms of case 2 I also agree, however I doubt it would need to change in the future, and if they were using the string match in a script it would be via the json format (I hope) and end up with an empty string which should be easier to check for than a string. Thoughts?

Reply:
actually the raw json should be `null` according to the opsman api docs, see [here](http://docs.pivotal.io/pivotalcf/2-3/opsman-api/#custom-ssl-certificate-settings). The empty string you got is because the type you used (string). If `interface{}` is used, then you would get the raw representation.

And for the second point, I would agree that it is not likely to change, but I would prefer to do it "right" if we can instead of assumption bad thing would not happen.

And to the question empty string, it depends on you. If you'd like to change it to interface, it would be an issue. But even empty issue would be easier to programmatically to handle than a hardcoded string. It is a stronger contract.